src/workspacer.Bar/BarSection.cs

```diff
@@ -38,85 +38,95 @@ public class BarSection
+            while (_panel.Controls.Count != _widgets.Count())
+            {
+                _panel.Controls.Add(CreateWidgetPannel());
+            }
+
             InitializeWidgets(widgets, _context);
         }
 
+        private FlowLayoutPanel CreateWidgetPannel()
+        {
+            return new FlowLayoutPanel
+            {
+                AutoSize = true,
+                Anchor = AnchorStyles.Top | AnchorStyles.Bottom,
+                BackColor = ColorToColor(_defaultBack),
+                Location = new Point(0, 0),
+                Margin = new Padding(0),
+                Size = new Size(50, 50),
+                WrapContents = false
+            };
+        }
+
         public void Draw()
         {
             if (_dirty)
             {
                 var widgets = _reverse ? _widgets.Reverse().ToArray() : _widgets;
-
-                int partNumber = 0;
-                for (var i = 0; i < widgets.Length; i++)
+                for (var wIndex = 0; wIndex < widgets.Length; wIndex++)
                 {
-                    var widget = widgets[i];
-                    var parts = widget.GetParts();
-                    for (var j = 0; j < parts.Length; j++)
+                    var widgetPanel = _panel.Controls[wIndex];
+                    var parts = widgets[wIndex].GetParts();
+
+                    EqualizeControls((FlowLayoutPanel)widgetPanel, parts.Count());
+
+                    for (var pIndex = 0; pIndex < parts.Length; pIndex++)
                     {
-                        var part = parts[j];
-
-                        if (partNumber < _panel.Controls.Count)
-                        {
-                            Label label = (Label)_panel.Controls[partNumber];
-                            SetLabel(label, part);
-                        }
-                        else
-                        {
-                            var label = AddLabel();
-                            SetLabel(label, part);
-                        }
-                        partNumber++;
+                        SetLabel((Label)widgetPanel.Controls[pIndex], parts[pIndex]);
                     }
                 }
 
-                var toRemove = new List<Control>();
-                if (partNumber < _panel.Controls.Count - 1)
+                _dirty = false;
+            }
+        }
+
+        private void EqualizeControls(FlowLayoutPanel pannel, int partCount)
+        {
+            if (pannel.Controls.Count != partCount)
+            {
+                while (pannel.Controls.Count < partCount)
+                {
+                    AddLabel(pannel);
+                }
+
+                while (pannel.Controls.Count > partCount)
                 {
-                    for (var i = partNumber; i < _panel.Controls.Count; i++)
-                    {
-                        toRemove.Add(_panel.Controls[i]);
-                    }
+                    pannel.Controls.RemoveAt(0);
```

Comment:
Wouldn't it be more efficient to have the removals occur at the end of the collection?